### PR TITLE
Use GPU with snapml

### PR DESCRIPTION
**Pull request by tanglef**
Same as LASSO: run snapml on GPU
| rcv1 low reg      | rcv1 high reg | news20 low reg  |
| ----------- | ----------- | ------------ |
| Capture d’écran de 2022-05-09 20-40-41| Capture d’écran de 2022-05-09 20-40-35 | Capture d’écran de 2022-05-09 20-40-23 |

### CODE REVIEW

**Comment by agramfort** [comment]
see my comment https://github.com/benchopt/benchmark_logreg_l1/pull/22/files#r868416503

File: solvers/snapml.py [hunk 1]
```diff
@@ -13,12 +16,14 @@ class Solver(BaseSolver):
...
     def set_objective(self, X, y, lmbd):
         self.X, self.y, self.lmbd = X, y, lmbd
+        self.use_gpu = cuda_version is not None
 
         self.clf = LogisticRegression(
             fit_intercept=False,
             regularizer=self.lmbd,
             penalty="l2",
             tol=1e-12,
+            use_gpu=self.use_gpu,
```
Comment (tomMoral) [inline solvers/snapml.py:26]: The recommendation in the [LogReg doc](https://snapml.readthedocs.io/en/latest/logistic_regression.html) is also to add:
```suggestion
            use_gpu=self.use_gpu,
            dual=self.X.shape[0] >= self.X.shape[1]
```

File: solvers/snapml.py [hunk 2]
```diff
@@ -13,12 +16,14 @@ class Solver(BaseSolver):
```
Comment (tomMoral) [inline solvers/snapml.py:16]:
```suggestion

    parameters = {
        'gpu': [True, Fase]
    }
    
    def skip(self, X, y, lmbd):
        if self.gpu and _get_cuda_version() is None:
            return True, "snapml[gpu] needs a GPU to run"
        return False, None

```

File: solvers/snapml.py [hunk 3]
```diff
@@ -1,4 +1,7 @@
 from benchopt import BaseSolver, safe_import_context
+from benchopt.utils.sys_info import _get_cuda_version
+
+cuda_version = _get_cuda_version()
```
Comment (tomMoral) [inline solvers/snapml.py:4]: I would only call this if the solver is being instanciated.
```suggestion
```

File: solvers/snapml.py [hunk 4]
```diff
@@ -13,12 +16,14 @@ class Solver(BaseSolver):
 
     def set_objective(self, X, y, lmbd):
         self.X, self.y, self.lmbd = X, y, lmbd
+        self.use_gpu = cuda_version is not None
```
Comment (tomMoral) [inline solvers/snapml.py:19]:
```suggestion
```

**Review by tomMoral — COMMENTED** [review]
Just a small question, are we sure the loss is not normalize? the first graph where the loss seems to converge to some other point is a bit weird no?

Otherwise, it is super helpful, thanks @tanglef, in particular for the plots !!

**Comment by tanglef** [comment]
> Just a small question, are we sure the loss is not normalize? the first graph where the loss seems to converge to some other point is a bit weird no?

I also wondered about that but:
- the loss used is described [here](https://ibmsoe.github.io/snap-ml-doc/v1.6.0/manual.html#logistic-regression) and it matches
- they even describe [here](https://ibmsoe.github.io/snap-ml-doc/v1.6.0/frequentlyaskedquestions.html#how-does-regularization-in-snap-ml-compare-to-sklearn) how to compare yourself against scikit-learn, and it matches again...
- we obtain almost the same sparsity in the coefficients (that are also quite close on `print`)

**Comment by tanglef** [comment]
Small question: how would I skip the tests for a solver's specific parameter (also relevant for LASSO and L1 logreg here)
The current error is:
` E       AssertionError: Solver skipped all simulated dataset configs. At least one simulated dataset config should be compatible with a solver
`
because we skip all runs when `gpu=True`

File: solvers/snapml.py [hunk 5]
```diff
@@ -11,6 +13,13 @@ class Solver(BaseSolver):
     install_cmd = "conda"
     requirements = ["pip:snapml"]
 
+    parameters = {"gpu": [True, False]}
```
Comment (tomMoral) [inline solvers/snapml.py:16]: Actually, the parameter per default of this solver would be the first one (implicitely, which is not great....)
So if you put this, the test should run without you needing to skip this:
```suggestion
    parameters = {"gpu": [False, True]}
```

Comment (tanglef) [reply solvers/snapml.py]: Thanks a lot! I'll use this for now but yes something more "intuitive" could be nicer here later

**Review by agramfort — APPROVED** [review]

**Comment by tomMoral** [comment]
Merging as this seems to be doing the right thing.
We can still open another issue if we see the solver not converge.

**Pull request merged**